#### BERT Architecture

##### BERT:   
bert stand for bidirectional encoder representations from transformers  
Text   
 ↓  
Tokenizer   
 ↓  
Tokens   
 ↓   
Embeddings  
 ↓   
Transformer Layers (BERT)   
 ↓   
Output

**Exercise 1:**

In [5]:
from transformers import AutoModel, AutoTokenizer

model_name = "bert-base-uncased"  #Loads BERT's tokenizer (converts text → token IDs)

tokenizer = AutoTokenizer.from_pretrained(model_name)   #Loads the full BERT model with all its weights (~110M parameters)

model = AutoModel.from_pretrained(model_name)

print(model)

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

[transformers] BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True, bias=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True, bias=True)
            (dropout): Dropout(p=

In [6]:
print(model.config)

BertConfig {
  "add_cross_attention": false,
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": null,
  "classifier_dropout": null,
  "dtype": "float32",
  "eos_token_id": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "is_decoder": false,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "tie_word_embeddings": true,
  "transformers_version": "5.12.1",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}



**Exercise 2**

In [7]:
text = "i love machine learning"

inputs = tokenizer(text, return_tensors = "pt")

print(inputs)

{'input_ids': tensor([[ 101, 1045, 2293, 3698, 4083,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1]])}


In [9]:
outputs = model(**inputs) #** unpacks a dictionary into keyword arguments

print(outputs.last_hidden_state.shape)

torch.Size([1, 6, 768])


**What Does This Mean?**   
1  
↓   
Batch size   
6   
↓  
Number of tokens   
768    
↓   
Features learned for each token   
BERT converted each token into a rich representation.

**Q1:** What problem does BERT solve compared to older NLP models?   
Answer:   
BERT understands the context and meaning of words by looking at surrounding words, whereas older NLP models often treated words independently and  
struggled with ambiguity.   

**Q2:** What is self-attention?    
Answer:    
Self-attention is a mechanism that allows a token to focus on and learn from other relevant tokens in the same sentence to better understand context   
and relationships.   

**Q3:** Why is BERT called bidirectional?    
Answer:    
BERT is called bidirectional because it reads text from both left-to-right and right-to-left simultaneously, allowing it to understand context from    both directions.  

**Q4:** What are transformer layers?    
Answer:     
Transformer layers are the building blocks of BERT that process embeddings using self-attention and neural networks to learn increasingly complex    
language patterns.   

**Q5:** What does torch.Size([1, 6, 768]) represent?   
Answer:   
1   → Batch size (1 sentence)   
6   → Number of tokens   
768 → Features learned for each token     

It means BERT produced a 768-dimensional representation for each of the 6 tokens in the sentence.